### PR TITLE
Support SSL connections to MongoDB

### PR DESCRIPTION
**Pull request by mossprescott**
This PR includes a small code change, adds a dependency, and most importantly includes a script that generates test certificates and details how to get it all to work.

@damonLL, please have a look and comment on the documentation here. Even better, pull the build and the script and see if you can get it to work by following the instructions there.

### CODE REVIEW

File: core/src/main/scala/quasar/physical/mongodb/util.scala [hunk 1]
```diff
@@ -88,14 +88,62 @@ object util {
...
+        .applyConnectionString(cs)
+        .invalidHostNameAllowed(invalidHostNameAllowed)
+        .build
+      settings.sslSettings(sslSettings)
+
+      // NB: Netty _must_ be used if SSL is required, but we do not use it by default
+      // mostly because it seems to cause the REPL to fail to exit cleanly. If necessary,
+      // it can also be forced using a system property (see the README).
+      if (sslSettings.isEnabled)
+        settings.streamFactoryFactory(new com.mongodb.connection.netty.NettyStreamFactoryFactory())
```
Comment (wemrysi) [inline core/src/main/scala/quasar/physical/mongodb/util.scala]: Mind wrapping this with braces since there is no `else` clause?

File: core/src/main/scala/quasar/physical/mongodb/util.scala [hunk 2]
```diff
@@ -88,14 +88,63 @@ object util {
...
+        .build)
+
+      settings.socketSettings(SocketSettings.builder
+        .applyConnectionString(cs)
+        .build)
+
+      val sslSettings = SslSettings.builder
+        .applyConnectionString(cs)
+        .invalidHostNameAllowed(invalidHostNameAllowed)
+        .build
```
Comment (drostron) [inline core/src/main/scala/quasar/physical/mongodb/util.scala]: `.build` could throw, but since we don't run on Java 1.6 seems ok to forgo wrapping in `Task`.

File: core/src/main/scala/quasar/repl/Main.scala [hunk 3]
```diff
@@ -137,7 +137,7 @@ object Main {
 
       coreApi      <- CoreEff.interpreter.liftM[MainErrT]
       ephemeralApi =  CfgsErrsIO.toMainTask(MntCfgsIO.ephemeral) compose coreApi
-      _            <- mountAll[CoreEff](config.mountings) foldMap ephemeralApi
+      _            <- (mountAll[CoreEff](config.mountings) foldMap ephemeralApi).flatMapF(_.point[Task])
```
Comment (drostron) [inline core/src/main/scala/quasar/repl/Main.scala]: Is this addition necessary?

Comment (mossprescott) [inline core/src/main/scala/quasar/repl/Main.scala]: Yes. What was happening there is `mountAll foldMap ephemeralApi` was returning `String \/ Unit`, and then we were binding that to `_`, discarding the error in the process. Flatmap/pointing puts the error on the left in MainTask, which gets it logged and stops forward progress, which is what we want for now in the case that mounting fails.

The following commit applies the same change to the api server.

File: core/src/main/scala/quasar/physical/mongodb/util.scala (hunk 2 above)
Comment (mossprescott) [inline core/src/main/scala/quasar/physical/mongodb/util.scala]: Wow, that's thorough. :smile: 

I think this code is arguably effectful if only because it does `new NettyStreamFactoryFactory()`, so I'm gonna go ahead and wrap it in Task. Note: That won't at the moment change the behavior anyway because it was already being _called_ from within Task.delay, but I think it makes sense to make it clear and avoid future problems.

File: core/src/main/scala/quasar/repl/Main.scala (hunk 3 above)
Comment (drostron) [inline core/src/main/scala/quasar/repl/Main.scala]: Makes sense. Thanks for the explanation.

File: core/src/main/scala/quasar/physical/mongodb/util.scala (hunk 2 above)
Comment (jdegoes) [inline core/src/main/scala/quasar/physical/mongodb/util.scala]: Again, it doesn't matter *where code is being called from*. If it's effectful, it *must* be wrapped in `Task` in order to maintain referential transparency.

The default assumption with any Java code should be that (a) it's effectful, and (b) it may return `null`, unless a detailed analysis of its documentation reveals stronger guarantees.

**Comment by drostron** [comment]
👍🏾, once @damonLL signs off.

**Comment by mossprescott** [comment]
Merged via the script
